aptos-move/block-executor/src/executor.rs

```diff
@@ -283,11 +294,51 @@ where
                         // Iterate round robin over workers to do commit_hook.
                         idx = (idx + 1) % post_commit_txs.len();
 
+                        // Committed the last transaction, BlockSTM finishes execution.
                         if txn_idx as usize + 1 == block.len() {
-                            // Committed the last transaction / everything.
                             scheduler_task = SchedulerTask::Done;
+
+                            counters::PARALLEL_PER_BLOCK_GAS.observe(accumulated_gas as f64);
                             break;
                         }
+
+                        // For committed txns with Success status, calculate the accumulated gas.
+                        // For committed txns with Abort or SkipRest status, early halt BlockSTM.
+                        match last_input_output.success_gas(txn_idx) {
+                            Ok(gas) => {
+                                accumulated_gas += gas;
+                                counters::PARALLEL_PER_TXN_GAS.observe(gas as f64);
+                            },
+                            _ => {
+                                scheduler.halt();
+
+                                counters::PARALLEL_PER_BLOCK_GAS.observe(accumulated_gas as f64);
+                                debug!("[BlockSTM]: Early halted due to Abort or SkipRest txn.");
+                                break;
+                            },
+                        };
+
+                        if let Some(per_block_gas_limit) = self.maybe_gas_limit {
+                            // When the accumulated gas of the committed txns exceeds PER_BLOCK_GAS_LIMIT, early halt BlockSTM.
+                            if accumulated_gas >= per_block_gas_limit {
+                                // Set the execution output status to be SkipRest, to skip the rest of the txns.
+                                last_input_output.update_to_skip_rest(txn_idx);
```

Comment:
why do we need this? isn't it halted anyway?

Reply:
Need to update the txn status to SkipRest to retry the rest of the txns.